api/v1alpha1/hostedcluster_types.go

```diff
@@ -78,6 +79,13 @@ type HostedClusterSpec struct {
 	// +kubebuilder:validation:Optional
 	// +kubebuilder:default={managementType: "Managed"}
 	Etcd EtcdSpec `json:"etcd"`
+
+	// Configs is a set of global config resources as defined in the
+	// openshift configuration API:
+	// https://docs.openshift.com/container-platform/4.7/rest_api/config_apis/config-apis-index.html
+	// Each entry contains a resource kind and corresponding configuration content.
+	// +kubebuilder:validation:Optional
+	Configs []ClusterConfiguration `json:"configs,omitempty"`
```

Comment:
Some questions:

1. Is every kind supported, or just a subset? 
2. Is the array validated to ensure the set is unique by kind?
3. How does this compare to an approach using a field per supported config kind (which would resolve 1 and 2 implicitly)?

Reply:
1. A subset
2. We should, a validating webhook is needed (which we've said we need for other validations as well)
3. The problem is that in the Hypershift API, which can be used for more than one control plane version, we can't assume to know what configs are supported for a potentially future release of OpenShift. If we introduce another global config type, our API is outdated.

Reply:
>The problem is that in the Hypershift API, which can be used for more than one control plane version, we can't assume to know what configs are supported for a potentially future release of OpenShift. If we introduce another global config type, our API is outdated.

Not sure I follow. I thought whatever is persisted at time N-1 needs to be backwards compatible with N; I'm not sure how the persisted representation (a field on the top level struct vs. an entry in the array) relates.

Maybe you could give a quick concrete example of how it breaks down from your perspective?

Reply:
So for example, with the knowledge of the latest release (4.8) we add a set of N fields for the kinds that we know we want to support today. 
The hypershift operator is deployed and goes into production. Let's say in 4.9 we introduce a new feature that requires its own config CRD. This is backwards compatible. However, our hypershift API will not allow this new feature to be configured because all we know about are the configs that existed in 4.8. The 4.9 version of our control plane operator would know about it, but the main hypershift API wouldn't.

Reply:
> So for example, with the knowledge of the latest release (4.8) we add a set of N fields for the kinds that we know we want to support today.
> The hypershift operator is deployed and goes into production. Let's say in 4.9 we introduce a new feature that requires its own config CRD. This is backwards compatible. However, our hypershift API will not allow this new feature to be configured because all we know about are the configs that existed in 4.8. The 4.9 version of our control plane operator would know about it, but the main hypershift API wouldn't.

And if we used an array, we would still need to wire support for dealing with that new type on the backend in the operator, right? Prior to that, the new config type just wouldn't be dealt with, which is okay, given backwards compatibility. In your scenario, the solution is to add the new field to the HyperShift API as part of establishing compatibility with 4.9.

Still not clear to me there's any issue either way in that regard. So which one provides the better API for the end user?

Reply:
> And if we used an array, we would still need to wire support for dealing with that new type on the backend in the operator, right?

That is the operator (control plane operator) that will be versioned with the release, so it is expected to have release-specific knowledge. We're not intending to version the hypershift API or the operator with each release.

Reply:
I don't think we expect our customers to upgrade hypershift with every new release of OCP.

Reply:
Given that, now I'm wondering how validation would work even in theory. Would what you're saying imply the webhook backend would be based on the specific release payload which is linked to the OCP API version embedded in the HostedCluster API raw field?

Reply:
No, what we can validate is that we're not duplicating entries of the same kind. That's it.
The control plane operator is responsible for interpreting the config and reporting whether its content is invalid.

Reply:
Alright, I think that answers all my questions, thanks